### PR TITLE
Wrap GMT's data structure GMT_GRID_HEADER for grid/image/cube headers

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Part of PR #2398 to support wrapping `GMT_GRID`/`GMT_IMAGE`/`GMT_CUBE`.

`GMT_GRID_HEADRE` is used in `GMT_GRID`/`GMT_IMAGE`/`GMT_CUBE`, so better to have it merged into the main branch then we can work on `GMT_GRID`/`GMT_IMAGE` in parallel.

Proposed by https://github.com/GenericMappingTools/pygmt/pull/2398#pullrequestreview-1947793913

No tests are added.

### CODE REVIEW

File: pygmt/datatypes/grid.py [hunk 1]
```diff
@@ -3,6 +3,99 @@
 """
```
Comment (weiji14) [inline pygmt/datatypes/grid.py:3]: At L2, can you mention that this file includes the `GMT_GRID_HEADER` struct?

File: pygmt/datatypes/grid.py [hunk 2]
```diff
@@ -3,6 +3,99 @@
...
+GMT_GRID_REMARK_LEN160 = 160
+
+# GMT uses single-precision for grids by default, but can be built to use
+# double-precision. Currently, only single-precision is supported.
+gmt_grdfloat = ctp.c_float
+
+
+class _GMT_GRID_HEADER(ctp.Structure):  # noqa: N801
+    """
+    GMT grid header structure for holding a grid header.
```
Comment (weiji14) [inline pygmt/datatypes/grid.py:25]:
```suggestion
    GMT grid header structure for metadata about the grid.
```

File: pygmt/datatypes/grid.py [hunk 3]
```diff
@@ -3,6 +3,99 @@
...
+        ("registration", ctp.c_uint32),
+        # Minimum/maximum x and y coordinates
+        ("wesn", ctp.c_double * 4),
+        # Minimum z value
+        ("z_min", ctp.c_double),
+        # Maximum z value
+        ("z_max", ctp.c_double),
+        # x and y increments
+        ("inc", ctp.c_double * 2),
+        # Grid values must be multiplied by this
```
Comment (weiji14) [inline pygmt/datatypes/grid.py:43]:
```suggestion
        # Grid values must be multiplied by this factor
```

File: pygmt/datatypes/grid.py [hunk 4]
```diff
@@ -3,6 +3,99 @@
...
+        ("y_units", ctp.c_char * GMT_GRID_UNIT_LEN80),
+        # Grid value units, in the form "long_name [units]"
+        ("z_units", ctp.c_char * GMT_GRID_UNIT_LEN80),
+        # Name of data set
+        ("title", ctp.c_char * GMT_GRID_TITLE_LEN80),
+        # Name of generating command
+        ("command", ctp.c_char * GMT_GRID_COMMAND_LEN320),
+        # Comments for this data set
+        ("remark", ctp.c_char * GMT_GRID_REMARK_LEN160),
+        # Below are itmes used internally by GMT
```
Comment (weiji14) [inline pygmt/datatypes/grid.py:59]:
```suggestion
        # Below are items used internally by GMT
```

**Comment by weiji14** [comment]
Can you copy some of the nice statements about GMT_GRID_HEADER from https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982 in your original comment? Good for future reference.

**Review by weiji14 — APPROVED** [review]
> No tests are added.

Ok with this, since the tests will be added in #2398.

Pre-approving for now, but feel free or not to include the recommendations above.

**Comment by seisman** [comment]
> Can you copy some of the nice statements about `GMT_GRID_HEADER` from [#2398 (comment)](https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982) into the first comment above? Good for future reference.

I'm not sure about this. Most of the statements in https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982 are about how to map GMT_GRID_HEADER metadata from/to netCDF metadata. They may make no sense to images.

**Comment by seisman** [comment]
> Can you copy some of the nice statements about `GMT_GRID_HEADER` from [#2398 (comment)](https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982) into the first comment above? Good for future reference.

I'm not sure about this. Most of the statements in https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982 are about how to map GMT_GRID_HEADER metadata from/to netCDF metadata. They may make no sense to images.

**Comment by weiji14** [comment]
> > Can you copy some of the nice statements about `GMT_GRID_HEADER` from [#2398 (comment)](https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982) into the first comment above? Good for future reference.
> 
> I'm not sure about this. Most of the statements in [#2398 (comment)](https://github.com/GenericMappingTools/pygmt/pull/2398#issue-1610245982) are about how to map GMT_GRID_HEADER metadata from/to netCDF metadata. They may make no sense to images.

I meant the `GMT_GRID_HEADER` data structure part, but that's ok, we can just leave it there.

**Pull request merged**